### DISCUSSION TITLE
RLS not letting me parse my own data

### DISCUSSION THREAD

**Discussion by alka7ex** [supabase/supabase#20298] [Questions] [answered]
Hi Guys

I try to parse my own data from sveltekit, previously i disable the rls and there is no issues, but when i want to implement rls now i can't access it. (and i already implement auth in sveltekit app also)

here is my code in sveltekit:

```
//+page.ts
import type { PageLoad } from './$types';
import { supabaseClient } from '$lib/supabaseClient';

const pageSize = 10;
const pageNumber = 1;

// Calculate the offset based on the page size and page number
const offset = (pageNumber - 1) * pageSize;

export const load = (async () => {
    return await supabaseClient
    .from("profiles")
    .select("*")
    .range(offset, offset + pageSize - 1)
    .limit(pageSize);
}) satisfies PageLoad;

```

here is my rls:

```
CREATE POLICY "Allow user to read own profile"
  ON "public"."profiles"
  FOR SELECT
  USING (auth.uid() = id);
```

so basically my goals is to only let authenticated user see their own data. but at the moment event without authenticated is still not working

![image](https://github.com/supabase/supabase/assets/18338838/4881a070-c5a7-4794-9eb5-bc2be419146d)

**Comment by alt-glitch** [comment]
Hey @alka7ex!
Your RLS policy looks okay. I used it myself. Can you make sure you have a column named `id` in your table? I remember making a similar mistake where I didn't have an `id` column and so `auth.uid() = id` wasn't evaluated properly.

Alternatively, if you're on the hosted version of Supabase - can you use the "impersonate" feature to see if it's a table/RLS issue or not.

Change your role to an authenticated user as shown below:
<img width="497" alt="image" src="https://github.com/supabase/supabase/assets/52913345/f2198edc-c96e-408a-9be2-900499fc83da">

Let me know if this helps!

**Reply by alka7ex** [reply]
Hi @alt-glitch 

Thank you for answering my questions
so yes i have the id column and auth.user.id = public.profiles.id (so basically same id)
![image](https://github.com/supabase/supabase/assets/18338838/34333b11-3601-4cc5-9417-41e526cef7b2)

and for impersonate i checked it and the data is there
![image](https://github.com/supabase/supabase/assets/18338838/59bb364a-f7a3-4b9c-b484-9632c027cd5f)

**Reply by alt-glitch** [reply]
Hey!
So the DB and the RLS are working fine. Seems that the problem is on the sveltekit side, specifically the auth. Are you sure the auth is implemented properly? 
Could you share where you're making the `console.log` statement? Maybe verify that the authorization header is set properly. It could be that you're sending these requests as the `anon` role rather than the authenticated user.
<img width="743" alt="image" src="https://github.com/supabase/supabase/assets/52913345/6488ba49-6fe4-4e65-9f91-7f84c0b6664d">

**Reply by alka7ex** [reply]
Hi @alt-glitch 

Here is the screenshot for the headers:
![image](https://github.com/supabase/supabase/assets/18338838/a526a67b-735e-4123-abb1-40ef038fca30)
so the profile is as public. and here is the proof:

and im using api that supabase provided:
![image](https://github.com/supabase/supabase/assets/18338838/2812eb84-705f-4594-b25a-71e4630f64ff)


here is my code for:
src/routes/(auth)/user/+page.server.ts
`import type { PageServerLoad } from './$types';
import { redirect } from '@sveltejs/kit' 
import { supabaseClient } from '$lib/supabaseClient';

export const load: PageServerLoad = async ({ url, locals: { getSession } }) => {
    try {
        const session = await getSession();
        if (!session) {
            throw redirect(302, '/login');
        }

        const data = await supabaseClient
            .from('profiles')
            .select("*");

        return{
            data
        }
    } catch (error) {
        console.error('Error in page load:', error);
        throw error; // Rethrow the error to handle it elsewhere if needed
    }
};
`
src/routes/(auth)/user/+page svelte
`<script lang="ts">
	export let data
	console.log(data); // i try to fetch data first before adding other components
</scipt>`
src/routes/+layout.server.ts
`import type { LayoutServerLoad } from './$types';

export const load: LayoutServerLoad = async ({ locals: { getSession } }) => {
	return {
		session: await getSession()
	};
};
`
src/routes/+layout.ts
`import { PUBLIC_SUPABASE_ANON_KEY, PUBLIC_SUPABASE_URL } from '$env/static/public';
import type { LayoutLoad } from './$types';
import { createBrowserClient, isBrowser, parse } from '@supabase/ssr';

export const load: LayoutLoad = async ({ fetch, data, depends }) => {
	depends('supabase:auth');

	const supabase = createBrowserClient(PUBLIC_SUPABASE_URL, PUBLIC_SUPABASE_ANON_KEY, {
		global: {
			fetch
		},
		cookies: {
			get(key) {
				if (!isBrowser()) {
					return JSON.stringify(data.session);
				}

				const cookie = parse(document.cookie);
				return cookie[key];
			}
		}
	});

	const {
		data: { session }
	} = await supabase.auth.getSession();

	return { supabase, session };
};
`
src/routes/+layout.svelte
`    <script lang="ts">
	import type { LayoutData } from './$types';
	import Navbar from '$lib/home/Navbar.svelte';
	import '../app.pcss';
	import { invalidate } from '$app/navigation';
	import { onMount } from 'svelte';
	import { derived, writable } from 'svelte/store';
	import Sidebar from '$lib/home/Sidebar.svelte';
	import { ModeWatcher } from "mode-watcher";


	export let data;
	// console.log(data)

	$: ({ supabase, session } = data);

	let supabaseStore = writable<typeof supabase>();
	$: supabaseStore.set(supabase);

	// this is necessary to ensure that subscriptions to old supabase clients are cleaned up properly
	// when a new client is retrieved from the loader.
	let supabaseAuthStateChangeSubscriptionStore = derived(supabaseStore, ($supabaseStore, set) => {
		const {
			data: { subscription }
		} = $supabaseStore.auth.onAuthStateChange((event, _session) => {
			if (_session?.expires_at !== session?.expires_at) {
				invalidate('supabase:auth');
			}
		});
		set(subscription);
		return subscription.unsubscribe;
	});

	onMount(() => {
		/* eslint-disable-next-line @typescript-eslint/no-empty-function */
		return supabaseAuthStateChangeSubscriptionStore.subscribe(() => {});
	});
        // export let data: LayoutData;
    </script>

<svelte:head>
    <title>GroRental Dashboard</title>
</svelte:head>

<div class="bg-background h-screen">
	<ModeWatcher />
	<Navbar data={data}/>
	<Sidebar/>
	<slot />
</div>

`
src/hooks.server.ts
`import { PUBLIC_SUPABASE_URL, PUBLIC_SUPABASE_ANON_KEY } from '$env/static/public';
import { createServerClient } from '@supabase/ssr';
import { redirect, type Handle } from '@sveltejs/kit';
import { sequence } from '@sveltejs/kit/hooks';

const createSupabaseClient: Handle = async ({ event, resolve }) => {

	event.locals.supabase = createServerClient(PUBLIC_SUPABASE_URL, PUBLIC_SUPABASE_ANON_KEY, {
		cookies: {
			get: (key) => event.cookies.get(key),
			// NOTE: defaulting path to '/' here to support Sveltekit v2 which requires it to be
			// specified.
			set: (key, value, options) => {
				event.cookies.set(key, value, { path: '/', ...options });
			},
			remove: (key, options) => {
				event.cookies.delete(key, { path: '/', ...options });
			}
		}
	});

	/**
	 * a little helper that is written for convenience so that instead
	 * of calling `const { data: { session } } = await supabase.auth.getSession()`
	 * you just call this `await getSession()`
	 */
	event.locals.getSession = async () => {
		const {
			data: { session }
		} = await event.locals.supabase.auth.getSession();
		return session;
	};

	if (event.route.id?.startsWith('/(auth)')) {
		if (!(await event.locals.getSession()))
		throw redirect(302, '/login');
	}

	return resolve(event, {
		filterSerializedResponseHeaders(name) {
			// supabase needs the content-range header
			return name === 'content-range';
		}
	});

};

export const handle = sequence(createSupabaseClient);
`
src/app.d.ts
`// See https://kit.svelte.dev/docs/types#app

import type { Database } from '$lib/generated/supabase/types';
import type { SupabaseClient, Session as SupabaseSession, User } from '@supabase/supabase-js';

// for information about these interfaces
declare global {
	namespace App {
		// interface Error {}

		interface Locals {
			supabase: SupabaseClient<Database>;
			getSession(): Promise<SupabaseSession | null>;
			session: Session| null;
		}

		interface PageData {
			supabase: SupabaseClient<Database>;
			session: User | null;
		}
		// interface PageState {}
		// interface Platform {}
	}
}

export {};
`

**Reply by alt-glitch** [reply]
I can't help you with your Svelte code because I'm not familiar with the framework. 
Maybe this could help: https://supabase.com/docs/guides/auth/auth-helpers/sveltekit

But I can see that the API call is made with the auth token as the `anon` key, it makes sense you're not getting any results because the user doesn't seem to be logging in properly. The bug is there in how you handle saving and accessing user sessions.

If you do `await supabaseClient.auth.getSession()` and get `null` it means the supabase client is not using the authenticated user session. I tried to do it with the code you provided and I was getting `null`.

**Reply by alka7ex** [reply]
@alt-glitch sorry man, but where you find that i use `anon` key? sorry my questions might be a dumb one, because i was at impression that im using the `public `one.

**Reply by alt-glitch** [reply, marked as answer]
Hey @alka7ex 
`anon` and `public` keys are the same. Because of RLS only an authenticated key will be able to read the table as you wish.
Maybe going through [these docs](https://supabase.com/docs/guides/auth) would help?

**Reply by alka7ex** [reply]
Hi @alt-glitch yes actually now it is fixed.

the problem is on my code where i didn't parse the session, so that's why it is not showed
thank you for helping me before
